docs: Add few core-concepts (#3229)
inspektor-gadget/inspektor-gadget: docs/core-concepts/_index.md

---
title: Core concepts
description: >
  This section describes the core concepts of Inspektor Gadget.
weight: 30
---

### Gadgets

A Gadget is an [OCI image](https://opencontainers.org/) that includes one or more eBPF programs, metadata YAML file and, optionally, a WASM module for post processing. They are the central component of Inspektor Gadget's framework. As OCI images that use the same tooling as containers and share the same attributes; shareable, modular, etc. Gadgets are built using the `ig image build` command and run with the `ig run`command.

![image](../images/gadget.png)

> Inspektor Gadget provides a collection of "official-gadgets", (TODO: link), but they can be created by any user. If you are interested in developing your gadget, please check TODO.

Reviewer: ```suggestion
Inspektor Gadget provides a collection of "official gadgets", (TODO: link), but they can be created by any user. If you are interested in developing your gadget, please check TODO.
```

> [!WARNING]

> > Prior to v0.31.0, Inspektor Gadget shipped gadgets in the binary. As of v0.31.0 these ***built-in*** Gadgets are still available and work as before but their use is discourage as they will be deprecate at some point. We encourage users to use ***image-based*** Gadgets going forward as they provide more features and decouple the eBPF programs from the Inspektor Gadget release process.

Reviewer: ```suggestion
> Prior to v0.31.0, Inspektor Gadget only shipped gadgets embedded in its executable file. As of v0.31.0 these ***built-in*** Gadgets are still available and work as before, but their use is discouraged as they will be deprecated at some point. We encourage users to use ***image-based*** Gadgets going forward, as they provide more features and decouple the eBPF programs from the Inspektor Gadget release process.
```

### Operators

Operators are a key building block of Inspektor Gadget. They form a pipeline that processes the data coming from gadgets to enrich, filter and export it.

![image](../images/operators-pipeline.png)

> Some operator we have are:

Reviewer: ```suggestion
Some operators included in Inspektor Gadget are:
```

> - Kube Manager
> - Local Manager
> - Formatter
> - CLI
> - Otel Metrics

Reviewer: ```suggestion
- Kube Manager: enriches gadget data in Kubernetes environments
- Local Manager: enriches gadget data in pure container environments
- Formatter: reformats gadget data, e.g. integers to human readable timestamps
- CLI: outputs gadget data to the console in either human readable form or as JSON, YAML, etc.
- Otel Metrics: collects and exports gadget data as metrics (Prometheus compatible)
```

#### Enrichment

> The data that eBPF collects from the kernel includes no knowledge about Kubernetes, container runtimes or any other high-level user-space concepts. In order to relate this data to these high-level concepts and make the eBPF data immediately more understandable, Inspektor Gadget automatically uses kernel primitives such as mount namespaces, pids, etc. to infer and which high-level concepts they relate to; Kubernetes pods, container names, DNS names, etc. The process of augmenting the eBPF data with these high-level concepts is called enrichment.

Reviewer: ```suggestion
The data that eBPF collects from the kernel includes no knowledge about Kubernetes, container runtimes or any other high-level user-space concepts. In order to relate the data to these high-level concepts and make the it immediately more understandable, Inspektor Gadget automatically uses kernel primitives such as mount namespaces, pids or similar to infer which high-level concepts they relate to: Kubernetes pods, container names, DNS names, etc. The process of augmenting the eBPF data with these high-level concepts is called enrichment.
```

#### Filtering

> Enrichment flows the other way, too. Inspektor Gadget enables users to do high-performance in-kernel filtering by only referencing high-level concepts such as Kubernetes pods, container names, etc.; automatically translating these to the corresponding low-level kernel resources.

Reviewer: ```suggestion
Enrichment works in the other direction, too: Inspektor Gadget enables users to do high-performance in-kernel filtering by only referencing high-level concepts such as Kubernetes pods or container names and automatically translating these to the corresponding low-level kernel resources.
```

#### Exporting

> The data can be exported to different places like the Console, Prometheus, gRPC, etc.

Reviewer: ```suggestion
Gadget data can be exported to different consumers like the console, Prometheus, gRPC or other third party software.
```
